crates/adapters/src/controller/stats.rs

```diff
@@ -199,6 +204,138 @@ pub struct ControllerStatus {
+    kind: MetricKind,
+
+    /// Metric name.
+    key: KeyName,
+
+    /// Optional key-value pairs that provide additional metadata about this
+    /// metric.
+    labels: Vec<(MetricString, MetricString)>,
+
+    /// Unit of measure for this metric, if any.
+    unit: Option<MetricUnit>,
+
+    /// Optional natural language description of the metric.
+    description: Option<MetricString>,
+
+    /// The metric's value.
+    value: MetricValue,
+}
+
+impl
+    From<(
+        CompositeKey,
+        Option<MetricUnit>,
+        Option<MetricString>,
+        MetricValue,
+    )> for ControllerMetric
+{
+    fn from(
+        (composite_key, unit, description, value): (
+            CompositeKey,
+            Option<MetricUnit>,
+            Option<MetricString>,
+            MetricValue,
+        ),
+    ) -> Self {
+        let (kind, key) = composite_key.into_parts();
+        let (name, labels) = key.into_parts();
+        Self {
+            kind,
+            key: name,
+            labels: labels.into_iter().map(|label| label.into_parts()).collect(),
+            unit,
+            description,
+            value,
+        }
+    }
+}
+
+impl Serialize for ControllerMetric {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let len = 3
+            + !self.labels.is_empty() as usize
+            + self.unit.is_some() as usize
+            + self.description.is_some() as usize;
+        let mut ts = serializer.serialize_struct("ControllerMetric", len)?;
+
+        let kind = match self.kind {
+            MetricKind::Counter => "counter",
+            MetricKind::Gauge => "gauge",
+            MetricKind::Histogram => "histogram",
+        };
+        ts.serialize_field("kind", kind)?;
+
+        ts.serialize_field("key", self.key.as_str())?;
+
+        if !self.labels.is_empty() {
+            // serde knows how to serialize `std::borrow::Cow` but `serde`
+            // defines and uses its own `Cow` that it can't handle.
+            let labels: Vec<_> = self
+                .labels
+                .iter()
+                .map(|(k, v)| (k.as_ref(), v.as_ref()))
+                .collect();
+            ts.serialize_field("labels", &labels)?;
+        } else {
+            ts.skip_field("labels")?;
+        }
+
+        if let Some(unit) = self.unit {
+            ts.serialize_field("unit", unit.as_canonical_label())?;
+        } else {
+            ts.skip_field("unit")?;
+        }
+
+        if let Some(description) = &self.description {
+            ts.serialize_field("description", description.as_ref())?;
+        } else {
+            ts.skip_field("description")?;
+        }
+
+        match &self.value {
+            MetricValue::Counter(counter) => ts.serialize_field("value", counter)?,
+            MetricValue::Gauge(gauge) => ts.serialize_field("value", &gauge.0)?,
+            MetricValue::Histogram(histogram) => {
```

Comment:
Not sure, maybe it's better to just report some quantiles here instead of the raw values